File: extract/extract_kafka.py
# extract_kafka.py
"""
Чтение сырых данных из Kafka (поток) и сохранение в HDFS.
Запуск:
  spark-submit --packages org.apache.spark:spark-sql-kafka-0-10_2.12:3.2.1 extract_kafka.py
"""
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql import functions as F

# Конфигурация
KAFKA_BOOTSTRAP = "kafka1:19092"          # внутренний listener
TOPICS = "amur_topic,neva_topic,ob_topic,volga_topic,yenisei_topic"
HDFS_OUTPUT_PATH = "hdfs://namenode:9000/raw/kafka_transactions"
CHECKPOINT_PATH = "hdfs://namenode:9000/checkpoints/kafka_raw"

spark = SparkSession.builder \
    .appName("Kafka Extract") \
    .config("spark.sql.streaming.forceDeleteTempCheckpointLocation", "true") \
    .getOrCreate()

# Читаем поток
raw_stream = spark.readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", KAFKA_BOOTSTRAP) \
    .option("subscribe", TOPICS) \
    .option("startingOffsets", "latest") \
    .load()

# Для сырого слоя сохраняем ключ, значение, топик, смещение, временную метку
raw_with_metadata = raw_stream.select(
    F.col("key").cast("string").alias("key"),
    F.col("value").cast("string").alias("value"),
    F.col("topic"),
    F.col("partition"),
    F.col("offset"),
    F.col("timestamp").alias("kafka_timestamp")
)

# Пишем в HDFS в формате Parquet
query = raw_with_metadata.writeStream \
    .format("parquet") \
    .option("path", HDFS_OUTPUT_PATH) \
    .option("checkpointLocation", CHECKPOINT_PATH) \
    .outputMode("append") \
    .trigger(processingTime="30 seconds") \
    .start()

print(f"Stream started, writing to {HDFS_OUTPUT_PATH}")
query.awaitTermination()

File: extract/ExtractKafka.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

object ExtractKafka {

  def main(args: Array[String]): Unit = {

    val spark = SparkSession.builder()
      .appName("Kafka Extract")
      .config("spark.sql.streaming.forceDeleteTempCheckpointLocation", "true") //Разрешает удалять временные checkpoint-файлы после завершения работы.
      .getOrCreate()

    val kafkaDF = spark.readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", "kafka1:19092") //Адрес Kafka Broker
      .option(
        "subscribe",
        "amur_topic,neva_topic,ob_topic,volga_topic,yenisei_topic"
      )
      .option("startingOffsets", "latest") //чтение только с новых сообщений
      .load()

    val rawDF = kafkaDF.select(
      col("key").cast("string").as("key"),
      col("value").cast("string").as("value"),
      col("topic"),
      col("partition"),
      col("offset"),
      col("timestamp").as("kafka_timestamp")
    )

    val query = rawDF.writeStream //настройка потоковой записи
      .format("parquet")
      .option("path", "hdfs://namenode:9000/raw/kafka_transactions")
      .option(
        "checkpointLocation",
        "hdfs://namenode:9000/checkpoints/kafka_raw"
      )
      .outputMode("append") //добавляет новые записи в существующие файлы
      .trigger(org.apache.spark.sql.streaming.Trigger.ProcessingTime("30 seconds")) //обработка выполняется каждые 30 секунд
      .start()

    println("Kafka stream started...")

    query.awaitTermination()
  }
}
